pm/validator_test.go

```diff
@@ -0,0 +1,167 @@
+package pm
+
+import (
+	"math/big"
+	"testing"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+func TestValidateTicket(t *testing.T) {
+	recipient := ethcommon.HexToAddress("73AEd7b5dEb30222fa896f399d46cC99c7BEe57F")
+	sender := ethcommon.HexToAddress("A69cdA26600c155cF2c150964Bdb5371ac3f606F")
+	sig := []byte("foo")
+	recipientRand := big.NewInt(10)
+	recipientRandHash := crypto.Keccak256Hash(ethcommon.LeftPadBytes(recipientRand.Bytes(), uint256Size))
+
+	b := newStubBroker()
+	sv := &stubSigVerifier{
+		shouldVerify: true,
+	}
+
+	v := NewValidator(recipient, b, sv)
+
+	// Test invalid recipient (null address)
+	ticket := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            ethcommon.Address{},
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		SenderNonce:       0,
+		RecipientRandHash: recipientRandHash,
+	}
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
+		t.Fatal("expected invalid recipient (null address)")
+	}
+
+	// Test invalid recipient (non-null address)
+	ticket = &Ticket{
+		Recipient:         sender,
+		Sender:            ethcommon.Address{},
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		SenderNonce:       0,
+		RecipientRandHash: recipientRandHash,
+	}
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
+		t.Fatal("expected invalid recipient (non-null address)")
+	}
+
+	// Test invalid sender
+	ticket = &Ticket{
+		Recipient:         recipient,
+		Sender:            ethcommon.Address{},
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		SenderNonce:       0,
+		RecipientRandHash: recipientRandHash,
+	}
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
+		t.Fatal("expected invalid sender")
+	}
+
+	// Test invalid preimage for commitment recipientRandHash
+	ticket = &Ticket{
+		Recipient:         recipient,
+		Sender:            sender,
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		SenderNonce:       0,
+		RecipientRandHash: ethcommon.Hash{},
+	}
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
+		t.Fatal("expected invalid preimage for commitment recipientRandHash")
+	}
+
+	// Test used ticket
+	ticket = &Ticket{
+		Recipient:         recipient,
+		Sender:            sender,
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		SenderNonce:       0,
+		RecipientRandHash: recipientRandHash,
+	}
+	// Set ticket as used in stub broker
+	b.RedeemWinningTicket(ticket, sig, recipientRand)
+
+	if err := v.ValidateTicket(ticket, sig, recipientRand); err == nil {
```

Comment:
General concern: shouldn't we validate that the error we're getting is the specific error we expect?